lenskit-core/src/main/java/org/lenskit/data/ratings/Rating.java

```diff
@@ -143,6 +169,51 @@ public double getValue() {
         return value;
     }
 
+    @Override
+    public Set<TypedName<?>> getTypedAttributeNames() {
+        return timestamp >= 0 ? FULL_ATTR_NAMES : NOTIME_ATTR_NAMES;
+    }
+
+    @Override
+    public boolean hasAttribute(String name) {
+        switch (name) {
+        case "id":
+        case "user":
+        case "item":
+        case "rating":
+            return true;
+        case "timestamp":
+            return timestamp > -1;
+        default:
+            return false;
+        }
+    }
+
+    @Nullable
+    @Override
+    public <T> T maybeGet(TypedName<T> name) {
+        return name.getType().cast(maybeGet(name.getName()));
```

Comment:
I think this will produce the wrong exception type if someone asks for (example) timestamp as a float. I think it would produce `ClassCastException` but it should produce `IllegalArgumentException`